### ISSUE TITLE
🚀 Feature: Core QueueService

### ISSUE THREAD

**Issue by drodil** [backstage/backstage#24116]
### 🔖 Feature description

Backstage core could provide a new service for queuing items and consuming these queues. The service should allow different kinds of queues like FIFO, FILO, etc. 

### 🎤 Context

`QueueService` could be used in multiple places like scheduled tasks to consume for example entities to process. Additional use-cases are welcome in this issue.

### ✌️ Possible Implementation

The basic implementation could be just an in-memory queue but this could be replaced to use for example Redis, Kafka, database, etc. 

### 👀 Have you spent some time to check if this feature request has been raised before?

- [X] I checked and didn't find similar issue

### 🏢 Have you read the Code of Conduct?

- [X] I have read the [Code of Conduct](https://github.com/backstage/backstage/blob/master/CODE_OF_CONDUCT.md)

### Are you willing to submit PR?

None

**Comment by dweber019** [comment]
A few days ago I almost implemented/used something like this, but my use case would add tasks, like a fetch request, to a queue and have a concurrency mechanism to limit the amount of tasks concurrently running.

This could be something to consider, maybe...

**Comment by freben** [comment]
@dweber019 did that queue have to be spread over multiple hosts? We use [p-limit](https://www.npmjs.com/package/p-limit) for local needs like that.

**Comment by drodil** [comment]
`p-limit` is nice for local queuing but having the possibility to split out work to multiple hosts would be beneficial. As the `CacheService` is already supporting Redis as one backend, maybe it could be used as the default implementation as it has queue support as well. Or even just `p-limit`, why not, cache also defaults to memory, right?

**Comment by andrewthauer** [comment]
I built something like this over top of [pgboss](https://github.com/timgit/pg-boss). Pairs nicely with the `TaskScheduler`, but could also replace it.

**Comment by Rugvip** [comment]
Had this on the radar for a while as something to add as a core service and I think it makes sense. It's still somewhat low prio though and we're waiting for more bespoke implementations in plugins. The ones have have right now in the catalog and scaffolder are very different so would be good to have some more input.

I do think an implementation on top of PostgreSQL is more or less always gonna be enough though, it's fast x)

**Comment by secustor** [comment]
I would like to see this too. For what it's worth I have implemented a QueueServices based for Redis and InMemory purposes which could be taken as inspiration. 

https://github.com/secustor/backstage-plugins/tree/main/plugins

Currently though there is no Knex option so it relies on a Redis cache to be available.

**Comment by andrewthauer** [comment]
> I would like to see this too. For what it's worth I have implemented a QueueServices based for Redis and InMemory purposes which could be taken as inspiration.
> 
> https://github.com/secustor/backstage-plugins/tree/main/plugins
> 
> Though, currently there is no Knex option so it relies on a Redis cache to be available.

Probably makes sense to abstract this so there can be different implementations. However, our preference for this would be a safe durable backing option. Having a job queue in memory is not ideal for the workloads we would likely use this for.

**Comment by secustor** [comment]
> Probably makes sense to abstract this so there can be different implementations. However, our preference for this would be a safe durable backing option. Having a job queue in memory is not ideal for the workloads we would likely use this for.

It is abstracted already, both the local and the Redis one are implementations of the same extension point. 
We use in Redis in production to coordinate between nodes and be persistent between restarts. The local in-memory queue on the other hand is simply for dev purposes.

**Comment by drodil** [comment]
@secustor any chance you could port these as core service to Backstage?

**Comment by secustor** [comment]
I don't think it will and should be added in the current form as a core service to Backstage. 
At least if not the maintainers are willing to add additional minimum requirements to Backstage. 

From my perspective I see these scenarios: 
1. Dev instance with SQLite --> I think here the implementation with an in memory queue is fine. 
2. Instance with PostgreSQL --> here we would need an implementation with something like the above mentioned `pgboss` 
3. Instance with other  SQL database --> My understanding is that Backstage officially supports all SQL databases which Knex supports too. In that case people would need to add a Redis instance or Backstage would have to provide their own queue implementation. 
4. Instance with Redis cache --> here we could use the Redis implementation. 

So to summarize there would be need a decision, if we want to force users to provide a Redis compatible cache, use Postgres with a newly written implementation or to write a complete own implementation of Queue only relying standard SQL functionality.

**Comment by freben** [comment]
> 3. Instance with other  SQL database --> My understanding is that Backstage officially supports all SQL databases which Knex supports too.

No, only pg, and additionally sqlite for tests and local dev. There's some best effort attempts at movement toward mysql in some chosen places/plugins, but no official promised support.

> 4. Instance with Redis cache --> here we could use the Redis implementation.

You aren't referring to using the existing `coreServices.cache` as a foundation right? Just wanted to highlight that that API is intentionally small and can't support a job queue properly. Maybe that's not what you meant.

> So to summarize there would be need a decision, if we want to force users to provide a Redis compatible cache, use Postgres with a newly written implementation or to write a complete own implementation of Queue only relying standard SQL functionality.

It'd be a high level interface with different underlying implementations, and I'd at least hope that the initial submission had a pg+sqlite implementation and then faster redis etc backing implementations could of course be added too.

**Comment by secustor** [comment]
> No, only pg, and additionally sqlite for tests and local dev. There's some best effort attempts at movement toward mysql in some chosen places/plugins, but no official promised support.

Good to know, that would enable a scenario for `pgboss` or `Redis` implementation for production use cases and in memory for dev purposes. 

> You aren't referring to using the existing coreServices.cache as a foundation right? Just wanted to highlight that that API is intentionally small and can't support a job queue properly. Maybe that's not what you meant.

No, my implementation simply reuses the Redis instance defined by cache, but does not use an APIs exposed by `coreServices.cache`. 

> It'd be a high level interface with different underlying implementations, and I'd at least hope that the initial submission had a pg+sqlite implementation and then faster redis etc backing implementations could of course be added too.
But it's worth mentioning that for example the events framework started out as in-memory-only and only recently a PR was added for a tentative pg implementation. So there are various possible sequencings!

Currently I've not an use case to implement it as the downstream project of the relevant plugin ( Renovate ) makes use of Redis anyway.
And so far none of my clients needed queues, but happy to contribute that feature if the opportunity comes up.

**Comment by drodil** [comment]
Not stale

**Comment by drodil** [comment]
Not stale
